Add documentation and guide for tracing schema (#12521)
mlflow/mlflow: docs/source/llms/tracing/tracing-schema.rst

              </p>
            </div>
            <div class="image-box">
                <img src="../../_static/images/llms/tracing/schema/trace_data_architecture.png"/>
            </div>
        </div>

.. |span-architecture| raw:: html

        <div class=""main-container"">
            <div>
              <h4>Span Structure</h4>
              <p> The Span object within MLflow's tracing feature provides detailed information about the individual steps of the trace. 
                Each Span object contains information about the step being instrumented, including the span_id, name, start_time, parent_id, status, 
                inputs, outputs, attributes, and events.
              </p>
            </div>
            <div class="image-box">
                <img src="../../_static/images/llms/tracing/schema/span_architecture.png"/>
            </div>
        </div>

.. container:: tracing-responsive-tabs

    .. tabs::

        .. tab:: Trace Architecture

            |trace-architecture|

        .. tab:: Trace Info

            |trace-info|

        .. tab:: Trace Data

            |trace-data|

        .. tab:: Span Architecture

            |span-architecture|

> Trace Schema

Reviewer: Will it be messy if we convert these sections as tables, probably one for each high-level object (TraceInfo / TraceData / Span)? The rational is that the h2 and h3 is hardly distinguishable in Sphinx, so it is not super easy to grasp the data structure from flat text sections. WDYT?

Author: Makes sense. It's a bit more dense, but I'm going with the table approach - it's a bit more concise

------------

A trace is composed of two components:

- :py:func:`mlflow.entities.trace_info.TraceInfo`

- :py:func:`mlflow.entities.trace_data.TraceData`

.. tip::
    Check the API documentation for helper methods on these dataclass objects for more information on how to convert or extract data from them.

Trace Info
----------

Trace Info is a dataclass object that contains metadata about the trace. This metadata includes information about the trace's origin, status, and 
various other data that aids in retrieving and filtering traces when used with :py:meth:`mlflow.client.MlflowClient.search_traces` and for 
navigation of traces within the MLflow UI.

To learn more about how ``TraceInfo`` metadata is used for searching, you can see examples :ref:`here <search_traces>`.

The data that is contained in the ``TraceInfo`` object is used to populate the trace view page within the MLflow tracking UI, as shown below.

.. figure:: ../../_static/images/llms/tracing/schema/trace_info_in_ui.png
    :alt: TraceInfo as it is used in the MLflow UI
    :width: 100%
    :align: center

> The primary components of MLflow :py:class:`mlflow.entities.trace_info.TraceInfo` objects are listed below.

Reviewer: ```suggestion
The primary components of MLflow :py:class:`TraceInfo <mlflow.entities.trace_info.TraceInfo>` objects are listed below.
```

Request ID
^^^^^^^^^^

The ``request_id`` of a trace is a unique identifier that is generated for each trace. This identifier is used within MLflow and integrated systems to 
resolve the event being captured and to provide associations for external systems to map the logged trace to the originating caller. 

.. note::

>     In the fluent API, this value will be generated for you. In the client API, you can provide a request_id to associate with the trace or omit 

Reviewer: This is a bit inaccurate. Regardless of which API they use, the request ID must be generated by MLflow (`MlflowClieit().start_trace()` generates request ID). What users can control is which generated ID they can pass to each spans.

Author: ahhh yeah. Fixed this

    this value to have it generated for you. If you require mapping trace data to an external system, it is recommended to provide a ``request_id`` that 
    is generated from your system to simplify the process of using the :py:meth:`mlflow.client.MlflowClient.search_traces` API for trace retrieval.

Experiment ID
^^^^^^^^^^^^^

The ``experiment_id`` property is a system-controlled immutable value that is used to associate the trace with the experiment in which it is logged. 
When searching through traces, this field is critical for filtering and grouping traces by experiment.

Timestamp (ms)
^^^^^^^^^^^^^^

The ``timestamp_ms`` property within TraceInfo marks the time at which the trace was created. This is a Unix timestamp in milliseconds.

.. note::

>     The time reflected here is the time at with the trace was created, not the time at which a request to your application was made. As such, 

Reviewer: +1 to have this noted in the doc!

    it does not factor into account the time it took to process the request to the environment in which your application is being served, which 
    may introduce additional latency to the total round trip time, depending on network configurations. 

Execution Time (ms)
^^^^^^^^^^^^^^^^^^^

> The ``execution_time_ms`` property within TraceInfo marks the total time that the trace took to complete. This is a Unix timestamp in milliseconds.

Reviewer 2: ```suggestion
The ``execution_time_ms`` property within ``TraceInfo`` marks the total time that the trace took to complete. This is a Unix timestamp in milliseconds.
```

Author: reworded to remove the reference

This time does not include the networking time associated with sending or receiving requests and responses from the environment that is interacting with 
the application.

Status
^^^^^^

A trace's status is represented by a ``TraceStatus`` enumeration object. The values that the status can have are:

- **OK** - The trace was successfully completed.
- **ERROR** - An error occurred during the trace event. Inspecting which span exhibited an error can aid in debugging.
- **IN_PROGRESS** - A trace has started and is awaiting completion. 
- **TRACE_STATUS_UNSPECIFIED** - The status of the trace is not set.

Request metadata
^^^^^^^^^^^^^^^^

The request metadata are additional key-value pairs of information that are associated with the Trace, set and modified by the tracing backend. 
These are not open for addition or modification by the user, but can provide additional context about the trace, such as an MLflow ``run_id`` that is 
associated with the trace. 

This metadata is immutable and considered system-controlled.

Tags
^^^^

> User-defined metadata that can be applied to a trace for applying additional context, aid in :ref:`search functionality <search_traces>`, or to 

Reviewer: ```suggestion
User-defined key-value pairs that can be applied to a trace for applying additional context, aid in :ref:`search functionality <search_traces>`, or to 
```

nit: cosmetic, but maybe we can avoid the term "metadata"

Attributes
^^^^^^^^^^

Attributes are metadata that are associated with a given step within your application. These attributes are key-value pairs that can be used to provide insight 
into behavioral modifications for function and method calls, giving insight into how modification of them can affect the performance of your application. 

Some common examples of attributes include:

- **model** - The name and version of a given external service when calling an LLM provider.
- **temperature** - The amount of creativity you are configuring for the answer from an LLM when calling it. 
- **document_count** - For vector store retrieval stages, the number of relevant documents to return to provide context to an LLM to answer a question.

Events
^^^^^^

Events are a system-level property that is optionally applied to a span only if there was an issue during the execution of the span. These events contain 
information about exceptions that were thrown in the instrumented call, as well as the stack trace. This data is structured within an 
:py:class:`mlflow.entities.SpanEvent` object.

> SpanEvent
> ~~~~~~~~~
> 
> The :py:class:`Span Event <mlflow.entities.SpanEvent>` object, if present, consists of three elements:
> 
> - **name** - defines the name of the event that occurred.
> - **timestamp** - the time at which the event occurred in microseconds.
> - **attributes** - a key-value collection of information related to the event that occurred. In the event of an exception being thrown, the stack trace 
>   will be included in this collection to aid in debugging and diagnosing the root cause of issues.

Reviewer: ```suggestion
```

nit: this seems to be repeating https://mlflow.org/docs/latest/python_api/mlflow.entities.html#mlflow.entities.SpanEvent (which is already linked)

Author: refactoring this
